libr/esil/esil_hooks.c

```diff
@@ -0,0 +1,363 @@
+/* radare2 - LGPL - Copyright 2023 - condret */
+
+#include <r_esil.h>
+#include <r_util.h>
+
+R_API REsilHooks *r_esil_hooks_new(void) {
+	REsilHooks *hooks = R_NEW0 (REsilHooks);
+	r_return_val_if_fail (hooks, NULL);
+	hooks->mem_read_observers = r_id_storage_new (0, UT32_MAX - 1);
+	hooks->mem_write_observers = r_id_storage_new (0, UT32_MAX - 1);
+	hooks->reg_read_observers = r_id_storage_new (0, UT32_MAX - 1);
+	hooks->reg_write_observers = r_id_storage_new (0, UT32_MAX - 1);
+	if (!(hooks->mem_read_observers && hooks->mem_write_observers &&
+		hooks->reg_read_observers && hooks->reg_write_observers)) {
+		r_id_storage_free (hooks->mem_read_observers);
+		r_id_storage_free (hooks->mem_write_observers);
+		r_id_storage_free (hooks->reg_read_observers);
+		r_id_storage_free (hooks->reg_write_observers);
+		free (hooks);
+		return NULL;
+	}
+	return hooks;
+}
+
+static bool free_hook_cb(void *user, void *data, ut32 id) {
+	free (data);
+	return true;
+}
+
+R_API void r_esil_hooks_free(REsilHooks *hooks) {
+	if (hooks) {
+		r_id_storage_foreach (hooks->mem_read_observers, free_hook_cb, NULL);
+		r_id_storage_foreach (hooks->mem_write_observers, free_hook_cb, NULL);
+		r_id_storage_foreach (hooks->reg_read_observers, free_hook_cb, NULL);
+		r_id_storage_foreach (hooks->reg_write_observers, free_hook_cb, NULL);
+	}
+	free (hooks);
+}
+
+R_API bool r_esil_set_mem_read_imp(REsil *esil, REsilImpHookMemReadCB imp, void *user) {
+	r_return_val_if_fail (esil && esil->hooks, false);
+	if (!esil->hooks->mem_read_implementation) {
+		esil->hooks->mem_read_implementation = R_NEW (REsilHook);
+	}
+	REsilHook *hook = esil->hooks->mem_read_implementation;
+	r_return_val_if_fail (hook, false);
+	hook->user = user;
+	hook->imr = imp;
+	return true;
+}
+
+R_API void r_esil_del_mem_read_imp(REsil *esil) {
+	r_return_if_fail (esil && esil->hooks);
+	R_FREE (esil->hooks->mem_read_implementation);
+}
+
+R_API bool r_esil_set_mem_write_imp(REsil *esil, REsilImpHookMemWriteCB imp, void *user) {
+	r_return_val_if_fail (esil && esil->hooks, false);
+	if (!esil->hooks->mem_write_implementation) {
+		esil->hooks->mem_write_implementation = R_NEW (REsilHook);
```

Comment:
uhm, only 1 handler per event? we cant have multiple listeners? that seems like not what we want right?